qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,349 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import Dict, Tuple
+
+import attrs
+import numpy as np
+from numpy.typing import ArrayLike
+
+from qualtran import Bloq, BloqBuilder, SelectionRegister, Signature, SoquetT
+from qualtran.bloqs.arithmetic import Add
+from qualtran.bloqs.basic_gates import OneEffect, OneState, ZeroEffect, ZeroState
+from qualtran.bloqs.basic_gates.rotation import Rx
+from qualtran.bloqs.qrom import QROM
+from qualtran.bloqs.select_and_prepare import PrepareOracle
+
+
+@attrs.frozen
+class ControlledStatePreparationUsingRotations(PrepareOracle):
+    r"""Class that implements controlled state preparation using Ry and Rz rotations from [1]. It
```

Comment:
nit: The class summary should be on line long (no line breaks) You can break it in two (separated by a blank line). "Class" is redundant here. Maybe just, Controlled state preparation using ....